src/Chainweb/Mempool/InMem.hs

```diff
@@ -510,42 +513,54 @@ insertCheckInMem' cfg lock txs
     curTxIdx <- withMVarMasked lock $ readIORef . _inmemCurrentTxs
 
     let withHashes :: Vector (T2 TransactionHash t)
-        withHashes = flip V.mapMaybe txs $ \tx ->
-          let !h = hasher tx
-          in (T2 h) <$> hush (validateOne cfg badmap curTxIdx now tx h)
+        withHashes = flip V.mapMaybe txs $ \(T2 h tx) ->
+          T2 h <$> hush (validateOne cfg badmap curTxIdx now tx h)
 
     V.mapMaybe hush <$!> _inmemPreInsertBatchChecks cfg withHashes
-  where
-    txcfg = _inmemTxCfg cfg
-    hasher = txHasher txcfg
 
 insertInMem
     :: forall t logger. (NFData t, Logger logger)
     => logger
     -> InMemConfig t    -- ^ in-memory config
+    -> IORef (HashSet TransactionHash)
     -> MVar (InMemoryMempoolData t)  -- ^ in-memory state
     -> InsertType
     -> Vector t  -- ^ new transactions
     -> IO ()
-insertInMem logger cfg lock runCheck txs0 = do
-    logFunctionText logger Debug $ "insertInMem: " <> sshow (runCheck, V.length txs0)
-    txhashes <- insertCheck
-    withMVarMasked lock $ \mdata -> do
-        pending <- readIORef (_inmemPending mdata)
-        logFunctionText logger Debug $ "insertInMem: pending txs: " <> sshow (HashMap.keys pending)
-        let cnt = HashMap.size pending
-        let txs = V.take (max 0 (maxNumPending - cnt)) txhashes
-        let T2 !pending' !newHashesDL = V.foldl' insOne (T2 pending id) txs
-        logFunctionText logger Debug $ "insertInMem: updated pending txs: " <> sshow (HashMap.keys pending')
-        let !newHashes = V.fromList $ newHashesDL []
-        writeIORef (_inmemPending mdata) $! force pending'
-        modifyIORef' (_inmemRecentLog mdata) $
-            recordRecentTransactions maxRecent newHashes
+insertInMem logger cfg pendingInsertionRef lock runCheck txs0 = do
+    pendingInsertionsDirty <- readIORef pendingInsertionRef
+    let
+      txs' = flip V.mapMaybe txs0 $ \tx ->
+        let hash = hasher tx
+        in if not (HashSet.member hash pendingInsertionsDirty)
+        then Just (T2 hash tx)
+        else Nothing
+    bracket
```

Comment:
@edmundnoble A bracket with functions that ignore their arguments is just `bracket_`.